app/lib/analyzer/scheduler.dart

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+/// Schedules and executes package analysis.
+class AnalysisScheduler {
+  final Stream taskReceivePort;
+  final List<ScheduledTask> _tasks = [];
+  final Set<ScheduledTask> _triggered = new Set();
+  StreamSubscription _subscription;
+  DateTime _lastDatastorePoll;
+
+  AnalysisScheduler(this.taskReceivePort);
+
+  Future run() async {
+    _subscription = taskReceivePort.listen((ScheduledTask task) {
+      _triggered.add(task);
+    });
+    await _runLoop();
+    // It is unlikely that we exit the run loop, but cancelling the subscription
+    // makes dartanalyzer happy.
+    await _subscription.cancel();
+  }
+
+  Future _runLoop() async {
+    for (;;) {
+      if (_tasks.isEmpty) {
+        await _populateTasks();
+        // short nap, just in case
+        await new Future.delayed(new Duration(seconds: 2));
+      } else {
+        final ScheduledTask task = _tasks.removeLast();
+        await _runAnalyze(task);
+        // long nap, to give GC ample time
+        await new Future.delayed(new Duration(seconds: 15));
+      }
+    }
+  }
+
+  Future _populateTasks() async {
+    final Set<ScheduledTask> newTasks = new Set();
+
+    // adding externally triggered tasks
+    if (_triggered.isNotEmpty) {
+      final List<ScheduledTask> triggeredList = new List.from(_triggered);
+      _triggered.clear();
+      for (ScheduledTask task in triggeredList) {
+        if (await _sanityCheck(task)) {
+          newTasks.add(task);
+        }
+      }
+    }
+
+    // polling the datastore every 10 minutes
+    final DateTime now = new DateTime.now();
+    if (_lastDatastorePoll == null ||
+        now.difference(_lastDatastorePoll).inMinutes >= 10) {
+      newTasks.addAll(await _pollDatastoreForTasks());
+      _lastDatastorePoll = new DateTime.now();
+    }
+
+    _tasks.addAll(newTasks);
+    _tasks.shuffle();
+  }
+
+  Future<bool> _sanityCheck(ScheduledTask task) async {
+    // TODO: check if the package and version exists
+    return false;
+  }
+
+  Future<List<ScheduledTask>> _pollDatastoreForTasks() async {
+    // TODO: get packages from datastore to analyze
+    return [];
+  }
+
+  Future _runAnalyze(ScheduledTask task) async {
+    // TODO: implement analysis
+  }
+}
```

Comment:
Since the search indexing might do the same thing: monitor new packages and index them (as well as going back in history), you could generalize this a bit.

For example something like this:

```dart

Future analysisIsolate() {
  await for (final Task task in getTasks()) {
    try {
       final analysis = await analyzePackage(task);
       await writeAnalysisResult(analysis);
     } catch (e, s) { ... }
  }
}

Future searchIndexingIsolate() {
  await for (final Task task in getTasks()) {
    try {
       final result = await indexPackage(task);
       index.update(result);
       await index.saveTo...();
     } catch (e, s) { ... }
  }
}

Stream<Task> getTasks() {
  // there is probably a `combineStreams` somewhere which does this.
  return combineStreams([
    new ManualTriggerSource().streamTasks(),
    new DatastoreHistorySource().streamTasks(),
    new DatastoreHeadPollerSource().streamTasks(),
  ]);
}

class TaskSource {
  Stream<Task> streamTasks();
}

class ManualTriggerSource extends TaskSource {
  ReceivePort port;

  Stream<Task> streamTasks() async* {
    await for (final message in port) {
      yield new Task(....);
    }
  }
}

class DatastoreHistorySource  extends TaskSource {
  Stream<Task> streamTasks() async* {
    await for (final candidate in dbService.query(PackageVersion, -date)) {
       yield new Task(candidate.package, candidate.version);
    }
  }
}

class DatastoreHeadPollerSource extends TaskSource {
  Stream<Task> streamTasks() async* {
    final DateTime begin = (await dbService.query(PackageVersion, -date, limit: 1)).first;
    while (true) {
      await new Future.delayed(const Duration(minutes: 1)); // polll interval.
      await for (final candidate in dbService.query(PackageVersion, -date)) {
        if (candidate.timeStamp > begin) yield candidate;
        else break;
      }
      yield new Task();
    }
  }
}

class Task { package, version, receivedTimestamp }
```